### ISSUE TITLE
Obj.truncate and Obj.set_field will be deprecated in OCaml 4.10

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#10602]
`Obj.truncate` and `Obj.set_tag` have been deprecated in OCaml upstream, c.f. https://github.com/ocaml/ocaml/pull/2279 and https://github.com/ocaml/ocaml/pull/2279

Users in Coq are `Hashset` (for `Obj.truncate`) and `Nativevalues` (for `Obj.set_tag`) , I am not sure how hard it will be to fix in the second case.

cc: @maximedenes @ppedrot

**Comment by ejgallego** [comment]
Adding a high priority tag as not solving this risk mis-compilation in future OCaml releases.

**Comment by ppedrot** [comment]
Fixing the hashset one is a matter of reusing the OCaml implementation for weak hashtables. As for the use in nativevalues, it's going to be much trickier, it's used to implement the accumulator in a very hackish way. I guess we can allocate it from C code instead.

**Comment by maximedenes** [comment]
> I am not sure how hard it will be to fix in the second case.

As far as `native_compute` is concerned, it should be easy to use the new `with_tag` primitive instead.

**Comment by ejgallego** [comment]
> As far as `native_compute` is concerned, it should be easy to use the new `with_tag` primitive instead.

What should the transition strategy be tho? Indeed that's a new function so we cannot just switch to it. It is feasible for us to provide an implementation of it for the moment?

**Comment by maximedenes** [comment]
> What should the transition strategy be tho?

We should probably follow what is recommended upstream. I would imagine staying with `set_tag` for now and switching when they remove it.

**Comment by ejgallego** [comment]
> We should probably follow what is recommended upstream. I would imagine staying with `set_tag` for now and switching when they remove it.

I understand the upstream strategy is more tricky than that, as they want to push multicore and this will mean we cannot use it, but that's from my reading of PR, we should ask.

Reimplementing `with_tag` seems the best path I think.

**Comment by maximedenes** [comment]
> Reimplementing `with_tag` seems the best path I think.

Are you sure it is possible? I thought they had implemented a specific optimization not to reallocate the value.

**Comment by ejgallego** [comment]
> Are you sure it is possible? I thought they had implemented a specific optimization not to reallocate the value.

Indeed the implemented version may not work the same? I am not familiar with that does the code in `simplif` do :/

**Comment by stedolan** [comment]
Is `Obj.set_tag` even needed in `nativevalues.ml`? As far as I can see, it's only used to set the tag of a newly-allocated record to `0`. But that's a no-op: records always have tag `0`. (I don't really understand what's going on in that file, so I'm probably missing something here)

**Comment by ppedrot** [comment]
@stedolan nope, it's used to set the tag of a *closure* to 0. Which is somewhat evil.

**Comment by stedolan** [comment]
Ah, I see. I missed a partial application in that file. That code is, um, exciting.

I think making a closure with tag other than Closure_tag is incompatible with no-naked-pointers mode anyway. Is it possible to do something else here? What's the goal of creating such a closure?

**Comment by ppedrot** [comment]
Yes, I am aware of the breakage of no-naked-pointers with the native compiler. This is essentially a hack that could be removed at the cost of a big refactoring of the compiler, potentially introducing a runtime slowdown.

Essentially, we want a notion of accumulator at the function type that can be observed. An invariant is that accumulators at every type must have tag 0 because they are somehow untyped.

**Comment by Alizter** [comment]
To make a note of this: If we switch to using malfunction for the native compiler, we wouldn't have to fight the OCaml compiler here. PMP and I discussed this previously and he says in order to do it the build system consolidation PR #15560  should be merged first.

Doing so will pave the way for OCaml 5.0 support.
